tavern/mqtt.py

```diff
@@ -0,0 +1,228 @@
+        check_expected_keys(expected_blocks["connect"], self._connect_args)
+
+        self._auth_args = kwargs.pop("auth", {})
+        check_expected_keys(expected_blocks["auth"], self._auth_args)
+
+        if "host" not in self._connect_args:
+            msg = "Need 'host' in 'connect' block for mqtt"
+            logger.error(msg)
+            raise exceptions.MissingKeysError(msg)
+
+        self._connect_timeout = self._connect_args.pop("timeout", 3)
+
+        # If there is any tls kwarg (including 'enable'), enable tls
+        self._tls_args = kwargs.pop("tls", {})
+        self._enable_tls = bool(self._tls_args)
+        # don't want to pass this through to tls_set
+        self._tls_args.pop("enable", None)
+
+        logger.debug("Paho client args: %s", self._client_args)
+        self._client = paho.Client(**self._client_args)
+        self._client.enable_logger()
+
+        if self._auth_args:
+            self._client.username_pw_set(**self._auth_args)
+
+        self._client.on_message = self._on_message
+
+        if self._enable_tls:
+            self._client.tls_set(**self._tls_args)
+
+        # Arbitrary number, could just be 1 and only accept 1 message per stages
+        # but we might want to raise an error if more than 1 message is received
+        # during a test stage.
+        self._message_queue = Queue(maxsize=10)
+        self._userdata = {
+            "queue": self._message_queue,
+        }
+        self._client.user_data_set(self._userdata)
+
+    @staticmethod
+    def _on_message(client, userdata, message):
+        """Add any messages received to the queue
+
+        Todo:
+            If the queue is faull trigger an error in main thread somehow
+        """
+        # pylint: disable=unused-argument
+
+        logger.info("Received mqtt message on %s", message.topic)
+
+        try:
+            userdata["queue"].put(message)
+        except Full:
+            logger.exception("message queue full")
+
+    def message_received(self, topic, payload, timeout=1):
+        """Check that a message is in the message queue
+
+        Args:
+            topic (str): topic message should have been on
+            payload (str, dict): expected payload - can be a str or a dict...?
+            timeout (int): How long to wait before signalling that the message
+                was not received.
+
+        Returns:
+            bool: whether the message was received within the timeout
+
+        Todo:
+            Allow regexes for topic names? Better validation for mqtt payloads
+        """
+
+        time_spent = 0
+
+        while time_spent < timeout:
+            t1 = time.time()
+
+            try:
+                msg = self._message_queue.get(block=True, timeout=timeout)
+            except Empty:
+                time_spent += timeout
+            else:
+                time_spent += time.time() - t1
+                if msg.payload != payload and msg.topic != topic:
```

Comment:
This if seems to fall into the `else` section if either the payload or the topic are correct, which does not appear to be the desired functionality. Looks that the and should be an `or`, or use the else block for the warning.

r.e. the comment, I don't think we should error on an unexpected message - I think we should just only pass when the exact right message is received.

Reply:
This should be fixed now